pairing_api/src/pairing.rs

```diff
@@ -0,0 +1,457 @@
+        &self,
+        topic: &str,
+        params: ResponseParamsSuccess,
+        message_id: MessageId,
+        client: &Client,
+    ) -> Result<(), PairingClientError> {
+        let irn_metadata = params.irn_metadata();
+        let response = Response::Success(SuccessfulResponse {
+            id: message_id,
+            jsonrpc: JSON_RPC_VERSION_STR.into(),
+            result: serde_json::to_value(params)
+                .map_err(|err| PairingClientError::EncodeError(err.to_string()))?,
+        });
+
+        // Publish the encrypted message
+        self.publish_payload(topic, irn_metadata, Payload::Response(response), client)
+            .await?;
+
+        Ok(())
+    }
+
+    /// Private function to publish a payload.
+    async fn publish_payload(
+        &self,
+        topic: &str,
+        irn_metadata: IrnMetadata,
+        payload: Payload,
+        client: &Client,
+    ) -> Result<(), PairingClientError> {
+        // try to extend session before updating local store.
+        let sym_key = {
+            let pairings = self.pairings.lock().await;
+            let pairing = pairings
+                .get(topic)
+                .ok_or_else(|| PairingClientError::PairingNotFound)?;
+            hex::decode(pairing.sym_key.clone()).map_err(|err| {
+                PairingClientError::EncodeError(format!("Failed to decode sym_key: {:?}", err))
+            })?
+        };
+
+        let payload = serde_json::to_string(&payload)
+            .map_err(|err| PairingClientError::EncodeError(err.to_string()))?;
+        let message = encrypt_and_encode(EnvelopeType::Type0, payload, &sym_key)
+            .map_err(|e| anyhow::anyhow!(e))
+            .map_err(|err| PairingClientError::EncodeError(err.to_string()))?;
+
+        // Publish the encrypted message
+        {
+            client
+                .publish(
+                    topic.into(),
+                    message,
+                    None,
+                    irn_metadata.tag,
+                    Duration::from_secs(irn_metadata.ttl),
+                    irn_metadata.prompt,
+                )
+                .await
+                .map_err(PairingClientError::PingError)?;
+        };
+
+        Ok(())
+    }
+
+    /// Private function to generate a WalletConnect URI.
+    fn generate_uri(pairing: &PairingInfo, sym_key: &str) -> String {
+        let mut url = format!(
+            "wc:{}@{}?symKey={}&relay-protocol={}&expiryTimestamp={}",
+            pairing.topic, VERSION, sym_key, pairing.relay.protocol, pairing.expiry
+        );
+
+        if !pairing.methods.0.is_empty() {
+            let methods_str = pairing
+                .methods
+                .0
+                .iter()
+                .map(|method_group| format!("[{}]", method_group.join(",")))
+                .collect::<Vec<_>>()
+                .join(",");
+
+            url.push_str(&format!("&methods={methods_str}"));
+        };
+
+        url
+    }
```

Comment:
it can be improved using less memory allocations

```rust
    fn generate_uri(pairing: &PairingInfo, sym_key: &str) -> String {
        let mut url = format!(
            "wc:{}@{}?symKey={}&relay-protocol={}&expiryTimestamp={}",
            pairing.topic, VERSION, sym_key, pairing.relay.protocol, pairing.expiry
        );

        if !pairing.methods.0.is_empty() {
            url.push_str("&methods=");

            pairing.methods.0.iter().for_each(|method_group| {
                url.push('[');
                url.push_str(&method_group.join(","));
                url.push(']');
                url.push(',');
            });

            // Remove trailing comma
            url.pop();
        }

        url
    }
```

here is the test

```rust
    #[test]
    fn test_generate_uri_with_methods() {
        let pairing = PairingInfo {
            topic: String::from("test_topic"),
            relay: Relay {
                protocol: String::from("test_relay_protocol"),
                data: None,
            },
            peer_metadata: None,
            expiry: 1234567890,
            active: true,
            methods: Methods(vec![
                vec![String::from("methodA"), String::from("methodB")],
                vec![String::from("methodC")],
            ]),
        };

        let sym_key = "test_sym_key";
        let expected_uri = "wc:test_topic@2?symKey=test_sym_key&\
                            relay-protocol=test_relay_protocol&expiryTimestamp=1234567890&\
                            methods=[methodA,methodB],[methodC]";

        let result = PairingClient::generate_uri(&pairing, sym_key);

        assert_eq!(result, expected_uri);
    }
```

please check it

Reply:
not really your version is manually constructing the methods string and removing the trailing comma which can make the logic more tedious and error-prone and also the memory overhead in my version isn't significant at all

Reply:
If you think that there wont be a lot of method groups and we can neglect the memory overhead then its fine